containers/attributes/hooks.py

```diff
@@ -0,0 +1,106 @@
+import uuid
+import json
+import datetime
+import os
+import logging
+import sys
+import functools
+from enum import Enum
+
+##### Set Up Logger ########
+
+AUDIT_LEVEL_NUM = os.getenv("AUDIT_LEVEL_NUM", 45)
+AUDIT_ENABLED = os.getenv("AUDIT_ENABLED", "false").lower() in ("yes", "true", "t", "1")
+
+logging.basicConfig(
+    stream=sys.stdout, level=os.getenv("SERVER_LOG_LEVEL", "CRITICAL").upper()
+)
+
+logging.addLevelName(AUDIT_LEVEL_NUM, "AUDIT")
+
+def audit(self, message, *args, **kws):
+    if self.isEnabledFor(AUDIT_LEVEL_NUM) and AUDIT_ENABLED:
+        self._log(AUDIT_LEVEL_NUM, message, args, **kws)
+
+logging.Logger.audit = audit
+logger = logging.getLogger(__package__)
+
+##### Custom Enums #########
+
+class CallType(Enum):
+    PRE = 1
+    POST = 2
+    ERR = 3
+
+class HttpMethod(Enum):
+    GET = 4
+    POST = 5
+    PUT = 6
+    PATCH = 7
+    DELETE = 8
+
+##### Hook decorator #########
+
+def run_with_hooks(http_method):
```

Comment:
Maybe just:
```suggestion
def hook(http_method):
```
or maybe `hook_into`